File: src/Service/Tren.java
package Service;

import EmployeeTasks.InterfazID;
import EmployeeTasks.Pila;

public abstract class Tren implements InterfazID{

    private String id;
    private int capacidadCarga;
    private int kilometraje;
    private Pila<Vagon> pilaVagones;

    public Tren(String id, int capacidadCarga, int kilometraje, int maximoVagones) {
        this.id = id;
        this.capacidadCarga = capacidadCarga;
        this.kilometraje = kilometraje;
        this.pilaVagones = new Pila<Vagon>(); 
    }

    @Override
    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public int getCapacidadCarga() {
        return capacidadCarga;
    }

    public void setCapacidadCarga(int capacidadCarga) {
        this.capacidadCarga = capacidadCarga;
    }

    public int getKilometraje() {
        return kilometraje;
    }

    public void setKilometraje(int kilometraje) {
        this.kilometraje = kilometraje;
    }

    public Pila<Vagon> getVagones() {
        return pilaVagones;
    }

    public void setPilaVagones(Pila<Vagon> pilaVagones) {
        this.pilaVagones = pilaVagones;
    }

    public void agregarVagon(Vagon vagon) {
        if (pilaVagones.getTamanio() >= pilaVagones.getCapacidad()) {
            System.out.println("La pila de vagones está llena.");
        } else {
            pilaVagones.push(vagon);
            System.out.println("Vagón " + vagon.getId() + " agregado al tren " + id + ".");
        }
    }

    public void eliminarVagon() {
        Vagon vagonEliminado = pilaVagones.pop();
        if (vagonEliminado != null) {
            System.out.println("Vagón " + vagonEliminado.getId() + " eliminado del tren " + id + ".");
        } else {
            System.out.println("No hay vagones para eliminar en el tren " + id + ".");
        }
    }

    public Vagon obtenerPrimerVagon() {
        return pilaVagones.peek();
    }

    public void mostrarVagones() {
        System.out.println("Vagones del tren " + id + ":");
        pilaVagones.mostrarElementos();
    }

    public abstract void verificarNumeroVagones();
}

File: Modulo3_administrartrenes/ClasesControladoras3/Tren.cs
using ProyectoEstructuras.Códigos.TareasEmpleado;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ProyectoEstructuras.Códigos.Servicio
{
    public abstract class Tren : InterfazID
    {
        private string nombre;
        private string id;
        private int capacidadVagones;
        private int kilometraje;
        private Pila<Vagon> pilaVagones;

        public string Nombre { get => nombre; set => nombre = value; }
        public string Id { get => id; set => id = value; }
        public int Kilometraje { get => kilometraje; set => kilometraje = value; }
        public Pila<Vagon> PilaVagones { get => pilaVagones; set => pilaVagones = value; }
        public int CapacidadVagones { get => capacidadVagones; set => capacidadVagones = value; }

        public Tren(string nombre, string id, int maximoVagones, int kilometraje)
        {
            this.nombre = nombre;
            this.id = id;
            this.capacidadVagones = maximoVagones;
            this.kilometraje = kilometraje;
            this.pilaVagones = new Pila<Vagon>();
        }

        public string GetId()
        {
            return Id;
        }

        public void AgregarVagon(Vagon vagon)
        {
            if (pilaVagones.Tamanio >= pilaVagones.Capacidad)
            {
                Console.WriteLine("La pila de vagones está llena.");
            }
            else
            {
                pilaVagones.Push(vagon);
                Console.WriteLine($"Vagón {vagon.GetId()} agregado al tren {id}.");
            }
        }

        public void EliminarVagon()
        {
            Vagon vagonEliminado = pilaVagones.Pop();
            if (vagonEliminado != null)
            {
                Console.WriteLine($"Vagón {vagonEliminado.GetId()} eliminado del tren {id}.");
            }
            else
            {
                Console.WriteLine($"No hay vagones para eliminar en el tren {id}.");
            }
        }

        public Vagon ObtenerPrimerVagon()
        {
            return pilaVagones.Peek();
        }

        public void MostrarVagones()
        {
            Console.WriteLine($"Vagones del tren {id}:");
            pilaVagones.MostrarElementos();
        }

        public abstract void VerificarNumeroVagones();
    }
}
